Implement safe-merge-queue

briannafan opened the pull request:
#### d3044ed093cd32ff02f5a4cc462e4e78f85820fd
<pre>
Implement safe-merge-queue
<a href="https://bugs.webkit.org/show_bug.cgi?id=261288">https://bugs.webkit.org/show_bug.cgi?id=261288</a>
rdar://114993838

Reviewed by Jonathan Bedard.

Adding the safe-merge-queue label adds the PR to a queue which is periodically checked through BuildBot.
If the PR passes EWS checks, merge-queue label is automatically added and the PR will be landed.
If the label is not added by a committer, or if the PR does not pass EWS checks, merging-blocked label is added.

* Tools/CISupport/ews-build/config.json:
* Tools/CISupport/ews-build/factories.py:
(UnsafeMergeQueueFactory.__init__):
(SafeMergeQueueFactory):
(SafeMergeQueueFactory.__init__):
* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps):
* Tools/CISupport/ews-build/loadConfig.py:
* Tools/CISupport/ews-build/steps.py:
(GitHubMixin):
(GitHubMixin.get_number_of_prs_with_label):
(ValidateCommitterAndReviewer):
(ValidateCommitterAndReviewer.fail_build):
(ValidateCommitterAndReviewer.run):
(DetermineLabelOwner.run):
(DetermineLabelOwner.getResultSummary):
(RemoveAndAddLabels):
(RemoveAndAddLabels.__init__):
(RemoveAndAddLabels.run):
(RemoveAndAddLabels.getResultSummary):
(RemoveAndAddLabels.update_labels):
(RetrievePRDataFromLabel):
(RetrievePRDataFromLabel.__init__):
(RetrievePRDataFromLabel.run):
(RetrievePRDataFromLabel.getResultSummary):
(RetrievePRDataFromLabel.getAllPRData):
(CheckStatusOfPR):
(CheckStatusOfPR.__init__):
(CheckStatusOfPR.run):
(CheckStatusOfPR.getResultSummary):
(CheckStatusOfPR.checkPRStatus):
(CheckStatusOfPR.getQueueStatusFromList):
(AddMergeLabelsToPRs):
(AddMergeLabelsToPRs.run):
(AddMergeLabelsToPRs.getResultSummary):
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/268847@main">https://commits.webkit.org/268847@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7dd0e8fd6b202b4e4e8ca412080060b1a02b0750

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20787 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21195 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21854 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22677 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19383 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21023 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24438 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21374 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20690 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21008 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20793 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18055 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23530 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17958 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18866 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25177 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19037 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19044 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23078 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19633 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16680 "Passed tests") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/20689 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18866 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23197 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2573 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19451 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/CISupport/ews-build/config.json:
```diff
@@ -392,6 +392,11 @@
+    {
+      "name": "Safe-Merge-Queue", "shortname": "safe-merge", "icon": "testOnly",
+      "factory": "SafeMergeQueueFactory", "platform": "*",
+      "workernames": ["ews151"]
```
Presumably this is just for testing, the final version should be run on the same machines tasked with `merge-queue`.

JonWBedard replied:
Since we're nearing landing this, it's time to switch this to the correct worker names (ie, our merge-queue bots)

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -145,6 +146,7 @@ def credentials(cls, user=None):
             return cls._cache[prefix]
 
         try:
+            print("i bet there's no passwords.json")
```
Nit: Remove debugging statement

briannafan replied:
Good catch ^^"

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2153,189 @@ def hideStepIf(self, results, step):
+        query = {'query': query_body}
+        num_prs = self.queryGraphQL(query)['data']['repository']['pullRequests']['totalCount']
+        if num_prs:
+            yield self._addToLog('stdio', 'There are {} PRs in safe-merge-queue.'.format(num_prs))
```
Every function with a `yield` should be decorated with `@defer.inlineCallbacks` and every function invoking a function with `@defer.inlineCallbacks` needs to preface that invocation of with `yield`

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2153,189 @@ def hideStepIf(self, results, step):
+            return {'step': f"Failed to request and filter pull requests"}
+        return buildstep.BuildStep.getResultSummary(self)
+
+    def queryGraphQL(self, payload):
```
I think this should be moved to the `GitHub` mixin class.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2153,189 @@ def hideStepIf(self, results, step):
+            else:
+                return FAILURE
+            self.setProperty('safe_pr_list', label_as_safe)
+            self._addToLog('stdio', 'PRs to merge: {}.'.format(label_as_safe))
```
`_addToLog` is asynchronous and should be prefaced with `yield`

JonWBedard commented on Tools/CISupport/ews-build/steps_unittest.py:
```diff
@@ -5746,6 +5746,105 @@ def test_skipped_branch(self):
+    def test_success_sort(self):
+        self.setupStep(FilterPullRequestsIntoQueues())
+        FilterPullRequestsIntoQueues.getNumPRs = lambda label: 4
+        query_result = {'data': {'repository': {'pullRequests': {'edges': [{'node': {'title': 'Fix `test-webkitpy webkitflaskpy`', 'number': 17412, 'commits': {'nodes': [{'commit': {'commitUrl': 'https://github.com/WebKit/WebKit/commit/582fb8b4f85cc9f385c0e0809170cadc48c7fed5', 'status': {'state': 'SUCCESS', 'contexts': [{'context': 'api-gtk', 'state': 'SUCCESS'}, {'context': 'api-ios', 'state':  ...
```
Can we organize this JSON a bit better? Maybe pretty-print it?

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2153,189 @@ def hideStepIf(self, results, step):
+        failed_status_check = []
+        no_checks = []
+
+        ignore_contexts = ['style', 'webkitperl', 'wpe', 'wpe-wk2', 'gtk', 'gtk-wk2', 'api-gtk', 'wincaire']
```
I think we want the reverse here: we want to list the tests we require to have passed before landing.

briannafan replied:
Sure, that would just change the proceeding logic a little. But, wouldn't it be preferable if a new test was added and required by default? In the required list implementation, I could see a test added to EWS but not this list and then a PR with an error going into merge-queue.

JonWBedard replied:
New EWS queues usually start as experimental. The bigger sticking point, though, is that's it's possible for a build to be pending. In that case, we don't want `safe-merge-queue` to be triggered (yet). The pending case is challenging because there may not be a context reported yet.

briannafan replied:
Okay, I'll make the change. I'm currently putting all PRs with no status into its own list that's ignored, perhaps that takes care of pending builds?

briannafan commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -193,6 +195,34 @@ def fetch_data_from_url_with_authentication_github(self, url):
         else:
             defer.returnValue(response)
 
+    @defer.inlineCallbacks
```
@JonWBedard Tried implementing twistd requests here and moved the function to GitHubMixin as requested!

aj062 commented on Tools/CISupport/ews-build/factories.py:
```diff
@@ -374,3 +374,10 @@ def __init__(self, platform, **kwargs):
+
+class SafeMergeQueueFactory(factory.BuildFactory):
+    def __init__(self, platform, configuration=None, architectures=None, additionalArguments=None, **kwargs):
+        super(SafeMergeQueueFactory, self).__init__()
```
Nit: can skip class name in super(). This code is Python 3 only.

JonWBedard replied:
If you meant to convert this to `super().__init__()`, @briannafan, that change has not been uploaded.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2174,188 @@ def hideStepIf(self, results, step):
+    @defer.inlineCallbacks
+    def run(self):
+        num_prs = yield self.getNumPRs()
+        if num_prs:
```
Can do early return here (so that you don't have to indent all the following code in this large if block). e.g.:

```
if not num_prs:
    defer.returnValue(FAILURE)
```

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2174,188 @@ def hideStepIf(self, results, step):
+        return buildstep.BuildStep.getResultSummary(self)
+
+    @defer.inlineCallbacks
+    def getNumPRs(self, label='\"safe-merge-queue\"'):
```
Can make the name more descriptive. e.g.: `getNumberOfPRsWithLabel`.
Also since this method is generic, it can be moved to base class `GitHubMixin`.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2174,188 @@ def hideStepIf(self, results, step):
+    @defer.inlineCallbacks
+    def run(self):
+        failed_prs = self.getProperty('blocked_pr_list', [])
+        pr_number = failed_prs.pop()
```
should check the length of list before calling pop()

briannafan replied:
Theoretically, this method would never run if `failed_status_check` is empty. But would it be better to include the check in case of some error?

aj062 replied:
It's still better to do the error check here, we don't want to blow up (fail the build with an exception) in case `blocked_pr_list` property is not set for some unexpected reason, or someone modifies the other step without realizing the dependency here.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2174,188 @@ def hideStepIf(self, results, step):
+
+    def getResultSummary(self):
+        if self.results == SUCCESS:
+            return {'step': f"Labelled blocked PRs with merge-blocked."}
```
Nit: this step seems to be labelling a single PR, not `PRs`. Can add the PR number in this message.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2174,188 @@ def hideStepIf(self, results, step):
+        if self.results == SUCCESS:
+            return {'step': f"Labelled blocked PRs with merge-blocked."}
+        elif self.results == FAILURE:
+            return {'step': f"Failed to label blocked PRs with merge-blocked."}
```
Ditto.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2174,188 @@ def hideStepIf(self, results, step):
+
+        did_remove_labels = yield self.remove_labels(pr_number, self.LABELS_TO_REMOVE, repository_url=repository_url)
+        if did_remove_labels:
+            yield self._addToLog('stdio', 'Successfully removed existing labels from PR {}.\n'.format(pr_number))
```
Nit: Can use f-strings here and through-out this code. Also better to mention here the name of labels removed (we are not removing all the existing labels).

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2174,188 @@ def hideStepIf(self, results, step):
         return not self.doStepIf(step)
 
 
+class FilterPullRequestsIntoQueues(buildstep.BuildStep, GitHubMixin, AddToLogMixin):
```
This name seems too generic and the name doesn't give much idea about what this class/step is excepted to do.

Can rename it to something like: `CheckStatusOfPRsInSafeMergeQueue` (and corresponding step name as well).

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2174,188 @@ def hideStepIf(self, results, step):
+            defer.returnValue(None)
+
+    @defer.inlineCallbacks
+    def sortPRs(self, num_prs, label='\"safe-merge-queue\"'):
```
what is this method doing? 
Doesn't look like it is doing exactly what the name says, which is sorting PRs. If it's getting list a of PRs, then the name should indicate so, e.g.: `getListOfPRsWithLabel` or `getSortedListOfPRsWithLabel` etc.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2174,188 @@ def hideStepIf(self, results, step):
+        passed_status_check = []
+        failed_status_check = []
+        no_checks = []
+        ios_tvos_watchos = ['ios', 'mac', 'ios-sim', 'ios-wk2', 'ios-wk2-wpt', 'api-ios', 'tv', 'tv-sim', 'watch', 'watch-sim']
```
We should move this list to more prominent place, atleast at the beginning of the class. Also make the variable name more descriptive. e.g.: QUEUES_FOR_SAFE_MERGE_QUEUE.

Also, I wonder if we should start with start a large list of queues, especially the testers queues like wk1, wk2, api tests) might have false positive. Can check if @JonWBedard @ryanhaddad have any strong opinion on it.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2150,6 +2174,188 @@ def hideStepIf(self, results, step):
+                no_checks.append(pr_num)
+                yield self._addToLog('stdio', 'No status. Done!\n')
+            else:
+                pr_commit_state = pr_commit_status_data['state']
```
You shouldn't retrieve EWS data about a commit from GitHub checks. We might stop sending that data there (e.g.: https://github.com/WebKit/WebKit/pull/4767).

Instead, get the data from ews.webkit.org, that service has holistic view of the PR/commit status and has detailed information about the PR/commit. You can use/modify `CheckStatusOnEWSQueues` step.

aj062 reviewed (commented):
Thanks for the PR. Good work. Few high-level notes:
- This code is Python 3 only. So, let's use python 3 features, especially f-string to improve readability.
- variable/method names can be improved, would be nice to have them descriptive, so that reading the name indicate what that method does.
- Let's get the EWS data about PR/commit from ews.webkit.org (see existing step CheckStatusOnEWSQueues).
- Please verify what all queues we want safe-merge-queue to block on to start-with.
- Would be nice if you can provide any sample URLs with this code in action (like on uat instance), would help in reviewing.

JonWBedard commented on Tools/CISupport/ews-build/loadConfig.py:
```diff
@@ -83,6 +83,7 @@ def loadBuilderConfig(c, is_test_mode_enabled=False, master_prefix_path=os.path.
         c['builders'].append(builder)
 
     c['prioritizeBuilders'] = prioritizeBuilders
+    # TODO: maybe add stuff here if anything
```
Can we remove this `TODO`?

briannafan replied:
Oops

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2255,6 +2269,305 @@ def hideStepIf(self, results, step):
 
 
+class RetrievePRDataFromLabel(buildstep.BuildStep, GitHubMixin, AddToLogMixin):
+    name = 'retrieve-pr-data-from-label'
```
Can we pass this to `build step.BuildStep` in `kwargs` instead? I ask because this should really be `retrieve-pr-data-from-label-<webkit/apple/webkit-security>`, so you basically want the name of the step to be changed based on the project.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2255,6 +2269,305 @@ def hideStepIf(self, results, step):
+        num_prs = yield self.get_number_of_prs_with_label(self.label)
+        if num_prs == 0:
+            yield self._addToLog('stdio', 'Ending process as there are no PRs in safe-merge-queue.\n')
+            return defer.returnValue(SUCCESS)
```
Inconsistent return behavior here. Would probably be most consistent to do something like this:
```
if num_prs == 0:
    yield self._addToLog('stdio', 'Ending process as there are no PRs in safe-merge-queue.\n')
    return defer.returnValue(SUCCESS)
if not num_prs:
    return defer.returnValue(FAILURE)
<rest of function>
```
but you could also remove this return statement and make everything an if/else.

JonWBedard replied:
This change was not included in the updated PR.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2255,6 +2269,305 @@ def hideStepIf(self, results, step):
+        return buildstep.BuildStep.getResultSummary(self)
+
+    @defer.inlineCallbacks
+    def getAllPRData(self, num_prs, label='\"safe-merge-queue\"'):
```
Can we change this argument to `limit`?

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2255,6 +2269,305 @@ def hideStepIf(self, results, step):
+        response = yield self.query_graph_ql(query)
+        if not response:
+            yield self._addToLog('stderr', 'Failed to retrieve list of PRs.\n')
+            defer.returnValue(None)
```
I think you want a `return` here.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2255,6 +2269,305 @@ def hideStepIf(self, results, step):
+
+
+class CheckStatusOfPRsInSafeMergeQueue(buildstep.BuildStep, GitHubMixin, AddToLogMixin):
+    name = 'check-status-of-prs-in-safe-merge-queue'
```
Can we again have this configured via kwargs? You should be able to have the name of this step be `check-status-of-pr-###-in-safe-merge-queue`, which would make things more clear.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2255,6 +2269,305 @@ def hideStepIf(self, results, step):
+        return buildstep.BuildStep.getResultSummary(self)
+
+
+class RemoveAndAddLabels(buildstep.BuildStep, GitHubMixin, AddToLogMixin):
```
Should we be consolidating with `BlockPullRequest` and `RemoveLabelsFromPullRequest`? Like previous steps (and especially if we're going to consolidate), we should probably be parameterizing `name`

JonWBedard commented on Tools/CISupport/ews-build/factories.py:
```diff
@@ -375,3 +375,11 @@ def __init__(self, platform, **kwargs):
+        super(SafeMergeQueueFactory, self).__init__()
+
+        self.addStep(RetrievePRDataFromLabel('WebKit/WebKit', label='\"safe-merge-queue\"'))
+        self.addStep(RetrievePRDataFromLabel('apple/WebKit', label='\"safe-merge-queue\"'))
```
Can we just iterate through all our remotes here? They should be stored in `GITHUB_PROJECTS`.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2255,6 +2271,310 @@ def hideStepIf(self, results, step):
+    flunkOnFailure = False
+    haltOnFailure = False
+
+    def __init__(self, label_to_add='', labels_to_remove=[GitHubMixin.SAFE_MERGE_QUEUE_LABEL], **kwargs):
```
You don't want to store a list in a default argument like this, it does weird things if you start modifying the list (since Python prefers to pass lists by reference instead of by value). Instead, you should do this:
```
def __init__(self, label_to_add='', labels_to_remove=None, **kwargs):
    self.label_to_add = label_to_add
    self.labels_to_remove = [GitHubMixin.SAFE_MERGE_QUEUE_LABEL] if labels_to_remove is None else labels_to_remove
    super().__init__(**kwargs)
```

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2255,6 +2271,310 @@ def hideStepIf(self, results, step):
+        else:
+            checked_status = yield self.checkPRStatus(pr_number)
+
+        # after all PRs have been filtered, add corresponding labels
```
No need for this comment.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2255,6 +2271,310 @@ def hideStepIf(self, results, step):
+        return buildstep.BuildStep.getResultSummary(self)
+
+    @defer.inlineCallbacks
+    def validateCommitterStatus(self, pr_number):
```
We're duplicating this code from `MergeQueue` and `UnsafeMergeQueue`, right? Is it possible for us to share it instead?

JonWBedard commented on Tools/CISupport/ews-build/factories.py:
```diff
@@ -375,3 +376,10 @@ def __init__(self, platform, **kwargs):
+    def __init__(self, platform, configuration=None, architectures=None, additionalArguments=None, **kwargs):
+        super().__init__()
+        for project in GITHUB_PROJECTS:
+            self.addStep(RetrievePRDataFromLabel(project, label='\"safe-merge-queue\"'))
```
Nit: I don't think we should be quoting the label name here. And the label name should probably be a variable (as our other label names names are). I do really like the fact that this label is passed as an argument to `RetrievePRDataFromLabel`, so let's be suer to keep that!

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -236,6 +236,20 @@ def query_graph_ql(self, payload):
             defer.returnValue(data)
 
+    @defer.inlineCallbacks
+    def get_number_of_prs_with_label(self, label='\"safe-merge-queue\"'):
```
This shouldn't have a default argument, since we only every call it by passing an argument (and I can't really see how future code would benefit from a default argument)

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2255,6 +2282,270 @@ def hideStepIf(self, results, step):
+    def __init__(self, project="WebKit/WebKit", label='\"safe-merge-queue\"', **kwargs):
+        self.project = project
+        owner, name = project.lower().split('/', 1)
+        self.name = f'retrieve-pr-data-from-{owner}-{name}'
```
This should be:
```
self.name = f'{self.name}-{owner}-{name}'
```
otherwise the class defined variable is completely useless.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2255,6 +2282,270 @@ def hideStepIf(self, results, step):
+    QUEUES_FOR_SAFE_MERGE_QUEUE = EMBEDDED_CHECKS + MACOS_CHECKS
+
+    def __init__(self, pr_number='', **kwargs):
+        self.name = f'check-status-of-pr-{pr_number}-in-safe-merge-queue'
```
Ditto on basing this on the class variable. Note that this does work in Python, and this is probably the right time to deploy this sort of tactic:
```
template_str = 'abc-{}-123'
resolved_str = template_str.format('def')
```

briannafan replied:
{} aren't allowed to be in Buildbot identifiers and makes unit tests fail. However, I think the step needs to be renamed anyways since it doesn't have to be safe-merge-queue specific, and it's only checking one PR, not multiple PRs. So now it's `CheckStatusOfPR()` with `self.name = '{self.name}-{pr_number}'`

rniwa reviewed (changes requested):
> Implementation of safe-merge-queue label through Buildbot that filters PRs through EWS checks into merge-queue and merging-blocked.

This description is still not self explanatory. Please use a full sentence to explain it.

JonWBedard reviewed (approved):
I want to collaborate with @briannafan on when to land and deploy, I am excited about this improvement!

briannafan committed:
Implement safe-merge-queue
https://bugs.webkit.org/show_bug.cgi?id=261288
rdar://114993838

Reviewed by Jonathan Bedard.

Adding the safe-merge-queue label adds the PR to a queue which is periodically checked through BuildBot.
If the PR passes EWS checks, merge-queue label is automatically added and the PR will be landed.
If the label is not added by a committer, or if the PR does not pass EWS checks, merging-blocked label is added.

* Tools/CISupport/ews-build/config.json:
* Tools/CISupport/ews-build/factories.py:
(UnsafeMergeQueueFactory.__init__):
(SafeMergeQueueFactory):
(SafeMergeQueueFactory.__init__):
* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps):
* Tools/CISupport/ews-build/loadConfig.py:
* Tools/CISupport/ews-build/steps.py:
(GitHubMixin):
(GitHubMixin.get_number_of_prs_with_label):
(ValidateCommitterAndReviewer):
(ValidateCommitterAndReviewer.fail_build):
(ValidateCommitterAndReviewer.run):
(DetermineLabelOwner.run):
(DetermineLabelOwner.getResultSummary):
(RemoveAndAddLabels):
(RemoveAndAddLabels.__init__):
(RemoveAndAddLabels.run):
(RemoveAndAddLabels.getResultSummary):
(RemoveAndAddLabels.update_labels):
(RetrievePRDataFromLabel):
(RetrievePRDataFromLabel.__init__):
(RetrievePRDataFromLabel.run):
(RetrievePRDataFromLabel.getResultSummary):
(RetrievePRDataFromLabel.getAllPRData):
(CheckStatusOfPR):
(CheckStatusOfPR.__init__):
(CheckStatusOfPR.run):
(CheckStatusOfPR.getResultSummary):
(CheckStatusOfPR.checkPRStatus):
(CheckStatusOfPR.getQueueStatusFromList):
(AddMergeLabelsToPRs):
(AddMergeLabelsToPRs.run):
(AddMergeLabelsToPRs.getResultSummary):
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/268847@main
